### ISSUE TITLE
Unit tests set up

### ISSUE THREAD

**Issue by James-Large** [uea-machine-learning/tsml#156]
**tldr**: have a look at the basic structure of these first-pass unit tests, lets decide on a format that people don't have objections to and then I'll fill out the same tests for some more classifiers before PR'ing into dev and allowing others to add their tests too

Trying to lay some basic foundations for minimal unit testing following recent issues on legacy and not-so-legacy code. 

Have pushed up branch [unitTestStuff](https://github.com/uea-machine-learning/tsml/tree/unitTestStuff), which contains the additions of some utility code for tests in ClassifierTools, a couple of examples (CAWPE and TSF) for basic reproduction testing and edited their mains to run the same. Test classes mirror the source code structure in src/test/java instead of src/main/java as is standard for testing 

Ideally, every classifier _at minimum_ should have an equivalent unit test to the ones in cawpe/tsf, or similar if in this issue we decide to change the format in some way. This test confirms that a) is can train and test b) the resulting test accuracy is at least the same as a previously confirmed value. This is obviously not amazingly thorough, but since italypowerdemand has a fair number of test cases, the classifier being 'incorrect' but achieving the same accuracy is unlikely.

For these, I edited the mains to have this form (stuff already there was commented out but retained) 

```
//        System.out.println(ClassifierTools.testUtils_getIPDAcc(new TSF(0)));
        System.out.println(ClassifierTools.testUtils_confirmIPDReproduction(new TSF(0), 0.967930029154519));
```

Where the value 0.967930029154519 was obtained by running the commented-out line

The unit test for TSF runs basically the same code

```
    @Test
    public void testReproduction() throws Exception {
        System.out.println("--testReproduction()");
        double expectedAcc = 0.967930029154519;
        System.out.println("TSF expected accuracy generated 2019_09_25");
        assertTrue(ClassifierTools.testUtils_confirmIPDReproduction(new TSF(0), expectedAcc));
    }
```

And when the entire project is tested in Netbeans (still for me...), it looks like this so far: 

![image](https://user-images.githubusercontent.com/44509982/65620307-7d8f5500-dfb9-11e9-867b-89abf84e9530.png)

Of course a generally nice thing in long term would be to have unit tests for as many things as possible. In the first pass, largely also because it should be easy for everyone to get on board with given the code provided, let's just limit it to classifiers.

**_Other details:_**

Have also played around a little with the travis and gradle settings. After much playing around with many things, I think the only change to these that has actually made it in is adding the --info option to .travis.yml line 8, the gradle test script call. This provides more information about the tests (as well as everything else, unfortunately), as without them it simply says 'x of y passed', and not which ones potentially failed.

Example travis build with intentionally failing test: https://travis-ci.com/uea-machine-learning/tsml/jobs/239084017

Example fully working travis build: https://travis-ci.com/uea-machine-learning/tsml/builds/129136231

On test outputs, stuff like adding 

```
test {
    testLogging {
        exceptionFormat = 'full'
    }
}
```

to build.gradle (recommended e.g [here](https://mrhaki.blogspot.com/2013/05/gradle-goodness-show-more-information.html) and [here](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.logging.TestLoggingContainer.html)) seems to error out since travis and/or gradle doesn't recognise it (the test tag, testlogging, etc all the way down). Presumably this is a versioning error somewhere. Not going to fix this right now, but playing around with that will likely be the answer to getting useful but concise outputs of the tests.

**Comment by goastler** [comment]
Cheers James, looks good. Are you using Junit, if so v4 or 5?

**Comment by James-Large** [comment]
Ah yeah that's something I forgot to mention, JUnit4 (4.12) that seems to get shipped with Netbeans and as was already being loaded by Gradle. So I haven't pushed up a standalone JUnit jar yet, but presumably will need to at some point. I'm curious whether IntelliJ will just pick it up 

There's utilities to run Junit3/4 tests in Junit5 ('vintage') if for some reason we decide to write a bunch in the former and update to the latter in the future, I haven't looked into the differences really. The main one I've seen is bigger/better tools for testsuites etc, i.e. defining test orderings/priors/environments more intelligently with 5 than 4

Nothing I've done so far should really break If we just do 5 from the get go. Having a quick look, seems like it's actually difficult to get a simple jar of junit5, they actively discourage it. junit 4.12 has jars available
